### PR TITLE
Update API References and ndarray documentation  

### PR DESCRIPTION
**Pull request by vlad-perevezentsev**
This PR updates API references to missing functions, refreshes `Multi-Dimensional Array(ndarray)` section and adds new sections `Array Indexing Routines` and `Data Type Routines` to `Routines`.

- [X] Have you provided a meaningful PR description?
- [ ] Have you added a test, reproducer or referred to issue with a reproducer?
- [ ] Have you tested your changes locally for CPU and GPU devices?
- [ ] Have you made sure that new changes do not introduce compiler warnings?
- [ ] Have you checked performance impact of proposed changes?
- [ ] If this PR is a work in progress, are you filing the PR as a draft?

### CODE REVIEW

**Comment by vlad-perevezentsev** [comment]
https://intelpython.github.io/dpnp/pull/1490/

**Review by antonwolfy — APPROVED** [review]
Great job! Thank you @vlad-perevezentsev, LGTM!

File: doc/reference/ndarray.rst [hunk 1]
```diff
@@ -12,3 +12,352 @@ For the basic concept of ``ndarray``\s, please refer to the `NumPy documentation
...
+
+   dpnp.ndarray
+
+
+Indexing arrays
+---------------
+
+Arrays can be indexed using an extended Python slicing syntax,
+``array[selection]``.  Similar syntax is also used for accessing
+fields in a :term:`structured data type`.
```
[the remark this answers is not in the document]
Comment (vlad-perevezentsev) [reply doc/reference/ndarray.rst:36]: No, this line shouldn't have been here. Thanks!

File: doc/reference/ndarray.rst [hunk 2]
```diff
@@ -12,3 +12,352 @@ For the basic concept of ``ndarray``\s, please refer to the `NumPy documentation
...
+For standard library functions:
+
+.. autosummary::
+   :toctree: generated/
+   :nosignatures:
+
+   dpnp.ndarray.__copy__
+   dpnp.ndarray.__deepcopy__
+   dpnp.ndarray.__reduce__
+   dpnp.ndarray.__setstate__
```
Comment (oleksandr-pavlyk) [inline doc/reference/ndarray.rst:320]: I do not think `cupy.ndarray` exposes these methods. What is the rationale for their support in `dpnp.ndarray`?

Comment (vlad-perevezentsev) [reply doc/reference/ndarray.rst:320]: These lines were added for future implementations and shouldn`t have been displayed in dpnp documentation.
I'll fix it. Thanks

**Pull request merged**